File: src/test/java/toby/command/commands/fetch/MemeCommandTest.java
package toby.command.commands.fetch;

import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.StatusLine;
import org.apache.http.client.HttpClient;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import toby.command.CommandContext;
import toby.command.commands.CommandTest;
import toby.dto.web.RedditAPIDto;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;

import static org.mockito.Mockito.*;

class MemeCommandTest implements CommandTest {

    MemeCommand memeCommand;

    String jsonResponse = "{\"data\": {" +
            "\"children\": [" +
            "{ " +
            "\"data\": {" +
            "\"title\": \"Sample Reddit Post\"," +
            "\"url\": \"https://www.old.reddit.com/r/raimimemes/sample-post\"," +
            "\"author\": \"sample_author\"," +
            "\"over_18\": false," +
            "\"is_video\": false" +
            "}" +
            "}" +
            "]" +
            "}" +
            "}";

    @BeforeEach
    void setUp() {
        setUpCommonMocks();
        doReturn(messageCreateAction)
                .when(messageChannelUnion)
                .sendMessageEmbeds(any(), any(MessageEmbed[].class));
        memeCommand = new MemeCommand();
    }

    @AfterEach
    void tearDown() {
        tearDownCommonMocks();
        reset(messageChannelUnion);

    }

    @Test
    void test_memeCommandWithSubreddit_createsAndSendsEmbed() throws IOException {
        //Arrange
        CommandContext commandContext = new CommandContext(event);
        OptionMapping subredditOptionMapping = mock(OptionMapping.class);
        HttpClient httpClient = mock(HttpClient.class);
        HttpResponse httpResponse = mock(HttpResponse.class);
        StatusLine statusLine = mock(StatusLine.class);
        HttpEntity httpEntity = mock(HttpEntity.class);
        InputStream contentStream = new ByteArrayInputStream(jsonResponse.getBytes());
        when(subredditOptionMapping.getAsString()).thenReturn("raimimemes");
        when(event.getOption("subreddit")).thenReturn(subredditOptionMapping);
        when(httpClient.execute(any())).thenReturn(httpResponse);
        when(httpResponse.getStatusLine()).thenReturn(statusLine);
        when(statusLine.getStatusCode()).thenReturn(200);
        when(httpResponse.getEntity()).thenReturn(httpEntity);
        when(httpEntity.getContent()).thenReturn(contentStream);

        //Act
        memeCommand.handle(commandContext, httpClient, requestingUserDto, 0);

        //Assert
        verify(interactionHook, times(1)).deleteOriginal();
        verify(messageChannelUnion, times(1)).sendMessageEmbeds(any(), any(MessageEmbed[].class));
    }

    @Test
    void test_memeCommandWithSubredditAndTimePeriod_createsAndSendsEmbed() throws IOException {
        //Arrange
        CommandContext commandContext = new CommandContext(event);
        OptionMapping subredditOptionMapping = mock(OptionMapping.class);
        OptionMapping timePeriodOptionMapping = mock(OptionMapping.class);
        HttpClient httpClient = mock(HttpClient.class);
        HttpResponse httpResponse = mock(HttpResponse.class);
        StatusLine statusLine = mock(StatusLine.class);
        HttpEntity httpEntity = mock(HttpEntity.class);
        InputStream contentStream = new ByteArrayInputStream(jsonResponse.getBytes());
        when(subredditOptionMapping.getAsString()).thenReturn("raimimemes");
        when(timePeriodOptionMapping.getAsString()).thenReturn(RedditAPIDto.TimePeriod.DAY.name().toUpperCase());
        when(event.getOption("subreddit")).thenReturn(subredditOptionMapping);
        when(event.getOption("timeperiod")).thenReturn(timePeriodOptionMapping);
        when(httpClient.execute(any())).thenReturn(httpResponse);
        when(httpResponse.getStatusLine()).thenReturn(statusLine);
        when(statusLine.getStatusCode()).thenReturn(200);
        when(httpResponse.getEntity()).thenReturn(httpEntity);
        when(httpEntity.getContent()).thenReturn(contentStream);

        //Act
        memeCommand.handle(commandContext, httpClient, requestingUserDto, 0);

        //Assert
        verify(interactionHook, times(1)).deleteOriginal();
        verify(messageChannelUnion, times(1)).sendMessageEmbeds(any(), any(MessageEmbed[].class));
    }
}

File: src/test/kotlin/toby/command/commands/fetch/MemeCommandTest.kt
package toby.command.commands.fetch

import io.mockk.*
import net.dv8tion.jda.api.interactions.commands.OptionMapping
import net.dv8tion.jda.api.interactions.commands.OptionType
import org.apache.http.HttpEntity
import org.apache.http.HttpResponse
import org.apache.http.StatusLine
import org.apache.http.client.HttpClient
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import toby.command.CommandContext
import toby.command.CommandTest
import toby.dto.web.RedditAPIDto
import java.io.ByteArrayInputStream
import java.io.IOException
import java.io.InputStream
import java.util.*

internal class MemeCommandTest : CommandTest {
    lateinit var memeCommand: MemeCommand

    private var jsonResponse: String = """{
        "data": {
            "children": [
                { 
                    "data": {
                        "title": "Sample Reddit Post",
                        "url": "https://www.old.reddit.com/r/raimimemes/sample-post",
                        "author": "sample_author",
                        "over_18": false,
                        "is_video": false
                    }
                }
            ]
        }
    }"""

    @BeforeEach
    fun setUp() {
        setUpCommonMocks()
        every { CommandTest.interactionHook.sendMessageEmbeds(any(), *anyVararg()) } returns CommandTest.webhookMessageCreateAction
        memeCommand = MemeCommand()
    }

    @AfterEach
    fun tearDown() {
        tearDownCommonMocks()
        clearMocks(CommandTest.messageChannelUnion)
    }

    @Test
    @Throws(IOException::class)
    fun test_memeCommandWithSubreddit_createsAndSendsEmbed() {
        //Arrange
        val commandContext = CommandContext(CommandTest.event)
        val subredditOptionMapping = mockk<OptionMapping>()
        val httpClient = mockk<HttpClient>()
        val httpResponse = mockk<HttpResponse>()
        val statusLine = mockk<StatusLine>()
        val httpEntity = mockk<HttpEntity>()
        val contentStream: InputStream = ByteArrayInputStream(jsonResponse.toByteArray())

        every { subredditOptionMapping.asString } returns "raimimemes"
        every { CommandTest.event.getOption("subreddit") } returns subredditOptionMapping
        every { httpClient.execute(any()) } returns httpResponse
        every { httpResponse.statusLine } returns statusLine
        every { statusLine.statusCode } returns 200
        every { httpResponse.entity } returns httpEntity
        every { httpEntity.content } returns contentStream

        //Act
        memeCommand.handle(commandContext, httpClient, CommandTest.requestingUserDto, 0)

        //Assert
        verify(exactly = 1) { CommandTest.interactionHook.sendMessageEmbeds(any(), *anyVararg()) }
    }

    @Test
    @Throws(IOException::class)
    fun test_memeCommandWithSubredditAndTimePeriod_createsAndSendsEmbed() {
        //Arrange
        val commandContext = CommandContext(CommandTest.event)
        val subredditOptionMapping = mockk<OptionMapping>()
        val timePeriodOptionMapping = mockk<OptionMapping>()
        val httpClient = mockk<HttpClient>()
        val httpResponse = mockk<HttpResponse>()
        val statusLine = mockk<StatusLine>()
        val httpEntity = mockk<HttpEntity>()
        val contentStream: InputStream = ByteArrayInputStream(jsonResponse.toByteArray())

        every { subredditOptionMapping.asString } returns "raimimemes"
        every { timePeriodOptionMapping.asString } returns RedditAPIDto.TimePeriod.DAY.name.uppercase(Locale.getDefault())
        every { CommandTest.event.getOption("subreddit") } returns subredditOptionMapping
        every { CommandTest.event.getOption("timeperiod") } returns timePeriodOptionMapping
        every { httpClient.execute(any()) } returns httpResponse
        every { httpResponse.statusLine } returns statusLine
        every { statusLine.statusCode } returns 200
        every { httpResponse.entity } returns httpEntity
        every { httpEntity.content } returns contentStream

        //Act
        memeCommand.handle(commandContext, httpClient, CommandTest.requestingUserDto, 0)

        //Assert
        verify(exactly = 1) { CommandTest.interactionHook.sendMessageEmbeds(any(), *anyVararg()) }
    }

    @Test
    fun testGetOptionData() {
        val optionDataList = memeCommand.optionData

        // Check the size of the returned list
        Assertions.assertEquals(3, optionDataList.size)

        // Check the properties of the OptionData objects
        val subreddit = optionDataList[0]
        val timePeriod = optionDataList[1]
        val limit = optionDataList[2]

        Assertions.assertEquals(OptionType.STRING, subreddit.type)
        Assertions.assertEquals("Which subreddit to pull the meme from", subreddit.description)
        Assertions.assertTrue(subreddit.isRequired)

        Assertions.assertEquals(OptionType.STRING, timePeriod.type)
        Assertions.assertEquals(
            "What time period filter to apply to the subreddit (e.g. day/week/month/all). Default day.",
            timePeriod.description
        )
        Assertions.assertFalse(timePeriod.isRequired)

        Assertions.assertEquals(OptionType.INTEGER, limit.type)
        Assertions.assertEquals("Pick from top X posts of that day. Default 5.", limit.description)
        Assertions.assertFalse(limit.isRequired)

        // Check the choices of the 'timePeriod' option
        val timePeriodChoices = timePeriod.choices
        Assertions.assertEquals(4, timePeriodChoices.size)
        Assertions.assertEquals("day", timePeriodChoices[0].name)
        Assertions.assertEquals("day", timePeriodChoices[0].asString)
        Assertions.assertEquals("week", timePeriodChoices[1].name)
        Assertions.assertEquals("week", timePeriodChoices[1].asString)
        Assertions.assertEquals("month", timePeriodChoices[2].name)
        Assertions.assertEquals("month", timePeriodChoices[2].asString)
        Assertions.assertEquals("all", timePeriodChoices[3].name)
        Assertions.assertEquals("all", timePeriodChoices[3].asString)
    }
}
